lib/email/receiver.rb

```diff
@@ -660,102 +660,97 @@ def trim_discourse_markers(reply)
       reply.split(reply_above_line_regex)[0]
     end
 
-    def parse_from_field(mail = nil)
-      mail ||= @mail
-
+    def parse_from_field(mail, process_forwarded_emails: true)
       if email_log.present?
         email = email_log.to_address || email_log.user&.email
         return email, email_log.user&.username
       elsif mail.bounced?
-        Array.wrap(mail.final_recipient).each { |from| return extract_from_address_and_name(from) }
+        # "Final-Recipient" has a specific format (<name> ; <address>)
+        # cf. https://www.ietf.org/rfc/rfc2298.html#section-3.2.4
+        address_type, generic_address = mail.final_recipient.to_s.split(";").map { _1.to_s.strip }
+        return generic_address, nil if generic_address.include?("@") && address_type == "rfc822"
       end
 
       return unless mail[:from]
 
       # For forwarded emails, where the from address matches a group incoming
       # email, we want to use the from address of the original email sender,
       # which we can extract from embedded_email_raw.
-      if has_been_forwarded?
-        if mail[:from].to_s =~ group_incoming_emails_regex && embedded_email[:from].errors.blank?
-          from_address, from_display_name = extract_from_fields_from_header(embedded_email, :from)
-          return from_address, from_display_name if from_address
+      if process_forwarded_emails && has_been_forwarded?
+        if mail[:from].to_s =~ group_incoming_emails_regex
+          if embedded_email && embedded_email[:from].errors.blank?
+            from_address, from_display_name =
+              Email::Receiver.extract_email_address_and_name(embedded_email[:from])
+            return from_address, from_display_name if from_address
+          end
         end
       end
 
+      # extract proper sender when using mailman mailing list
+      if mail[:x_mailman_version].present?
+        address, name = Email::Receiver.extract_email_address_and_name_from_mailman(mail)
+        return address, name if address
+      end
+
       # For now we are only using the Reply-To header if the email has
       # been forwarded via Google Groups, which is why we are checking the
       # X-Original-From header too. In future we may want to use the Reply-To
       # header in more cases.
-      if mail["X-Original-From"].present?
-        if mail[:reply_to] && mail[:reply_to].errors.blank?
-          from_address, from_display_name =
-            extract_from_fields_from_header(
-              mail,
-              :reply_to,
-              comparison_headers: ["X-Original-From"],
-            )
-          return from_address, from_display_name if from_address
-        end
-      end
-
-      if mail[:from].errors.blank?
-        from_address, from_display_name = extract_from_fields_from_header(mail, :from)
-        return from_address, from_display_name if from_address
+      if mail[:x_original_from].present? && mail[:reply_to].present?
```

Comment:
Refactored all our methods for extracting the email / name from the `From` header into just one method `extract_email_address_and_name` so I refactored that bit.